components/lock/consul/consul_lock.go

```diff
@@ -0,0 +1,192 @@
+package consul
+
+import (
+	"errors"
+	"github.com/hashicorp/consul/api"
+	"mosn.io/layotto/components/lock"
+	"mosn.io/pkg/log"
+	"strconv"
+	"sync"
+)
+
+const (
+	address       = "address"
+	scheme        = "scheme"
+	username      = "username"
+	password      = "password"
+	defaultScheme = "http"
+)
+
+type ConsulTxn interface {
+	Txn(txn api.TxnOps, q *api.QueryOptions) (bool, *api.TxnResponse, *api.QueryMeta, error)
+}
+type SessionFactory interface {
+	Create(se *api.SessionEntry, q *api.WriteOptions) (string, *api.WriteMeta, error)
+}
+
+type ConsulLock struct {
+	metadata       metadata
+	logger         log.ErrorLogger
+	client         *api.Client
+	sessionFactory SessionFactory
+	txn            ConsulTxn
+	mMap           map[string]lockMessage
+	mutex          sync.Mutex
+}
+
+type lockMessage struct {
+	LockOwner string
+	Session   string
+	Index     uint64
+}
+
+func NewConsulLock(logger log.ErrorLogger) *ConsulLock {
+	consulLock := &ConsulLock{
+		mMap: make(map[string]lockMessage, 0),
+	}
+	return consulLock
+
+}
+
+func (c *ConsulLock) Init(metadata lock.Metadata) error {
+	consulMetadata, err := parseConsulMetadata(metadata)
+
+	if err != nil {
+		return err
+	}
+	c.metadata = consulMetadata
+	client, err := api.NewClient(&api.Config{
+		Address: consulMetadata.address,
+		Scheme:  consulMetadata.scheme,
+	})
+	c.client = client
+	c.sessionFactory = client.Session()
+	c.txn = client.Txn()
+	return nil
+}
+func (c *ConsulLock) Features() []lock.Feature {
+	return nil
+}
+
+func getTTL(expire int32) string {
+	//session TTL must be between [10s=24h0m0s]
+	if expire < 10 {
+		expire = 10
+	}
+	return strconv.Itoa(int(expire)) + "s"
+}
+
+func (c *ConsulLock) TryLock(req *lock.TryLockRequest) (*lock.TryLockResponse, error) {
+
+	// create a session TTL
+	session, _, err := c.sessionFactory.Create(&api.SessionEntry{
+		TTL:       getTTL(req.Expire),
+		LockDelay: 0,
+	}, nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	//use transaction to lock
+	_, response, _, err := c.txn.Txn(api.TxnOps{
```

Comment:
这里我没太懂，为啥用事务呀？
我查了一下，好像Consul的session类似于etcd的lease，拿session去acquire key，就能实现分布式锁，似乎用不到事务?
https://blog.didispace.com/spring-cloud-consul-lock-and-semphore/
https://guidao.github.io/lock.html#org4fb7b5a

话说我花了好长时间才看明白consul咋用……这东西的文档洋洋洒洒写了那么多字 看着还是莫名其妙

Reply:
@seeflood 这里我用事务的原因是，事务可以返回response，里面带着kv的基本信息，例如: ModifyIndex ,  unlock的时候CAS删除节点需要这个ModifyIndex。 为什么不用release？ 直接用release的话，consul并不会删除kv，会不会造成很多无用的kv ？

Reply:
噢噢明白你意思了。我查了一下用session+release应该可以，具体来说：

- 如果想实现“session到期，自动删除key”，可以在创建session的时候[指定behavior](https://www.consul.io/api/session#session_create)
>Behavior (string: "release") - Controls the behavior to take when a session is invalidated. Valid values are:
release - causes any locks that are held to be released
delete - causes any locks that are held to be deleted

- 如果想实现unlock的时候cas，可以[在release的时候传sessionid](https://www.consul.io/api/kv#create-update-key)
>release (string: "") - Supply a session ID to use in a release operation. This is useful when paired with ?acquire= as it allows clients to yield a lock. This will leave the LockIndex unmodified but will clear the associated Session of the key. The key must be held by this session to be unlocked.

这样似乎不会删掉key,只是释放key的session

- 或者，unlock的时候直接删掉session好像就行，如果创建session的时候behavior传的是delete，那么session失效时consul会自动删掉Key
https://www.consul.io/api/session#delete-session

实现可以参考[这篇博客](https://blog.didispace.com/spring-cloud-consul-lock-and-semphore/)

Reply:
至于事务的实现有没有缺点，我没细看，要查一下。用事务会有性能损失么？